packages/react-events/src/FocusScope.js

```diff
@@ -0,0 +1,157 @@
+  context: ReactResponderContext,
+  state: FocusScopeState,
+): void {
+  const elements = context.getFocusableElementsInScope();
+  if (elements.length > 0) {
+    const firstElement = elements[0];
+    firstElement.focus();
+    state.currentFocusedNode = firstElement;
+  }
+}
+
+function focusLastChildEventTarget(
+  context: ReactResponderContext,
+  state: FocusScopeState,
+): void {
+  const elements = context.getFocusableElementsInScope();
+  const length = elements.length;
+  if (elements.length > 0) {
+    const lastElement = elements[length - 1];
+    lastElement.focus();
+    state.currentFocusedNode = lastElement;
+  }
+}
+
+const FocusScopeResponder = {
+  rootEventTypes,
+  createInitialState(): FocusScopeState {
+    return {
+      nodeToRestore: null,
+      currentFocusedNode: null,
+    };
+  },
+  onRootEvent(
+    event: ReactResponderEvent,
+    context: ReactResponderContext,
+    props: FocusScopeProps,
+    state: FocusScopeState,
+  ) {
+    const {type, target, nativeEvent} = event;
+
+    if (type === 'focus') {
+      if (context.isTargetWithinEventComponent(target)) {
+        state.currentFocusedNode = ((target: any): HTMLElement);
+      } else if (props.trap) {
+        if (state.currentFocusedNode !== null) {
+          state.currentFocusedNode.focus();
+        } else {
+          focusFirstChildEventTarget(context, state);
+        }
+      }
+    } else if (type === 'keydown' && nativeEvent.key === 'Tab') {
+      const currentFocusedNode = state.currentFocusedNode;
+      if (currentFocusedNode !== null) {
+        const {altkey, ctrlKey, metaKey, shiftKey} = (nativeEvent: any);
+        // Skip if any of these keys are being pressed
+        if (altkey || ctrlKey || metaKey) {
+          return;
+        }
+        const elements = context.getFocusableElementsInScope();
+        const position = elements.indexOf(currentFocusedNode);
+        if (shiftKey) {
+          if (position === 0) {
+            if (props.trap) {
+              focusLastChildEventTarget(context, state);
+            } else {
+              return;
+            }
+          } else {
+            const previousElement = elements[position - 1];
+            previousElement.focus();
+            state.currentFocusedNode = previousElement;
+          }
+        } else {
+          if (position === elements.length - 1) {
+            if (props.trap) {
+              focusFirstChildEventTarget(context, state);
+            } else {
+              return;
+            }
+          } else {
+            const nextElement = elements[position + 1];
+            nextElement.focus();
+            state.currentFocusedNode = nextElement;
+          }
+        }
+        ((nativeEvent: any): KeyboardEvent).preventDefault();
+      }
+    }
+  },
+  onMount(
+    context: ReactResponderContext,
+    props: FocusScopeProps,
+    state: FocusScopeState,
+  ) {
+    if (props.restoreFocus) {
+      state.nodeToRestore = document.activeElement;
```

Comment:
For this component to work inside of iframes this should be the ownerDocument of the current node.

Reply:
Thanks for the feedback. I addressed these points in https://github.com/facebook/react/pull/15496 :)